sklearn/linear_model/_base.py

```diff
@@ -518,7 +526,29 @@ def fit(self, X, y, sample_weight=None):
             # Sample weight can be implemented via a simple rescaling.
             X, y = _rescale_data(X, y, sample_weight)
 
-        if sp.issparse(X):
+        if self.solver == "cholesky":
+            n_samples, n_features = X.shape
+            ravel = False
+            if y.ndim == 1:
+                y = y.reshape(-1, 1)
+                ravel = True
+            n_samples_, n_targets = y.shape
+            alpha = np.asarray(0, dtype=X.dtype).ravel()
+
+            if n_targets > 1:
+                alpha = np.repeat(alpha, n_targets)
+
+            try:
+                self.coef_ = _cholesky_helper(X, y, alpha, n_features,
+                                              n_samples)
+            except TypeError:
+                raise TypeError('X matrix is singular and sparse, and not'
```

Comment:
I understand the docs to say svd will be adopted automatically if the data is singular. Perhaps the docs are not quite right? I also think it would be better to explicitly check for sparse data and raise an appropriate error, rather than just responding to a TypeError.

Reply:
@jnothman Given some input, if it is non-singular, then Cholesky can handle it. If it is singular, then it uses the SVD solver. But the SVD solver does not work if the matrix is also sparse. 

The `TypeError` that is being caught here is only raised by `_cholesky_helper` when the matrix is explicitly found to be sparse. 

Maybe a more elegant solution is to explicitly check when a matrix is singular and kick such queries to the `elif sp.issparse(X) ...` code block instead of the Cholesky solver, but I think it's computationally costly to check for singularity. Thoughts?

Reply:
Yes, I don't think we should check for singularity.

Reply:
Why can't we do what's done in Ridge (I haven't checked what's done there).

Reply:
@amueller Ridge currently catches `LinAlgError` (e.g. singular matrix) and kicks it to the SVD solver. The SVD solver returns a `TypeError` on sparse input. So I think this is essentially equivalent to what happens in Ridge.